### ISSUE TITLE
ECR IAM Role usage

### ISSUE THREAD

**Issue by spaghettifunk** [keel-hq/keel#425]
I tried using the `iam role` in the pod annotation to avoid putting credentials in the `values.yaml` file and push it to GitHub but it seems that keel is not able to pick up the role correctly. 
I was wondering if somebody else managed to make it work with `podsAnnotations` rather than plain credentials. If not, I'll be happy to work on a PR 😃

**Comment by severity1** [comment]
I also want this, based on my tests it seems that even if you set `ecr.enabled: false` this just doesn't set the aws access, secret and region variables and values as seen below.
```
env:
        - name: NAMESPACE
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: metadata.namespace
        - name: POLL
          value: "1"
        - name: HELM_PROVIDER
          value: "1"
        - name: TILLER_NAMESPACE
          value: kube-system
        - name: TILLER_ADDRESS
        - name: BASIC_AUTH_USER
          value: potatoman
        - name: SLACK_CHANNELS
          value: deployment
        - name: SLACK_APPROVALS_CHANNEL
          value: deployment
        - name: SLACK_BOT_NAME
          value: keelbot
        - name: NOTIFICATION_LEVEL
          value: info
```

tried this and keel can still detect that my deployment is using an ECR repo although its giving the usual 401 Not Authorized as seen below.
```
time="2019-08-13T04:48:52Z" level=error msg="trigger.poll.manager: got error(-s) while watching images" error="encountered errors while adding images: Get https://blahblah.dkr.ecr.ap-southeast-2.amazonaws.com/v2/banana/bloop/manifests/0.0.6: http: non-successful response (status=401 body=\"Not Authorized\\n\")"
```

this suggests to me that ecr.enabled just sets variables and values, to me this is a bit misleading. just make `aws.enabled`.

although this also suggests that even if it can still detect that you are using an ecr repo and goes through the ecr logic, something in there explicitly references to the aws key/val to be present.

my expectation is that, if AWS key/val is not present anywhere always fallback to IAM roles/profiles.

**Comment by rusenask** [comment]
@spaghettifunk I would really appreciate a PR for this! Could you extend https://github.com/keel-hq/keel/tree/master/extension/credentialshelper/aws package to include this additional logic?

**Comment by spaghettifunk** [comment]
@rusenask I'd be happy to! I will start working on it on Friday because we need to have the role implemented for our CI/CD system 😄

**Comment by RiceBowlJr** [comment]
Hi there!
I just configured Keel on AWS to poll on ECR.
I do not like to put credentials in any app conf, neighter generate a whole lot of user for apps.

I use Kiam to solve this issue and it works well with Keel :) 

If this can help, here is what to do:
* [Annotate](https://github.com/uswitch/kiam#overview) your namespace for Kiam.
* Create a role for Keel, assumable by the Kiam role, with the following policy:
```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Sid": "",
            "Effect": "Allow",
            "Action": [
                "ecr:ListImages",
                "ecr:GetRepositoryPolicy",
                "ecr:GetDownloadUrlForLayer",
                "ecr:DescribeRepositories",
                "ecr:DescribeImages",
                "ecr:BatchGetImage",
                "ecr:BatchCheckLayerAvailability"
            ],
            "Resource": [
                "arn:aws:ecr:<region>:<account_number>:repository/<repo-name>"
            ]
        },
        {
            "Sid": "",
            "Effect": "Allow",
            "Action": "ecr:GetAuthorizationToken",
            "Resource": "*"
        }
    ]
}
```
* Add the annotation to your Keel pods, in the manifest/Helm chart:
```
podAnnotations:
  iam.amazonaws.com/role: <role_arn>
```

And you are good to go :)

**Comment by spaghettifunk** [comment]
Ha! I must confess that I did not expect this. We are currently using `kube2iam` and keel was not able to pull from ECR. We tried even with Admin rights and still we were having `401 Permission Denied`.  I'll try again maybe I overlooked something. Thanks!

**Comment by severity1** [comment]
i've been using `kube2iam` as well, i wonder why it doesn't work.

**Comment by severity1** [comment]
@spaghettifunk it is quite possible that KIAM works because it populates the ENV VARS while kube2iam populates the credentials file or just plain IAM Instance Profile, if this is the case the underlying issue still stands that the ecr logic doesnt failover to the credentialhelper lookup and is being blocked by ENV VARS existence.

**Comment by spaghettifunk** [comment]
Yeah could be. I've never used `kiam` so I can't really know per-sei. On the other hand, I'm working actively onto a PR for this issue because we don't want to have AWS credentials exposed in our repository. Hopefully by the end of the week I'm able to open the PR with the changes.

**Comment by flythebluesky** [comment]
Also very interested in using Kube2IAM w/AWS ECR. If Keel uses the AWS Default Credentials Provider chain instead of only environment variables, it should work. We have the same concerns regarding exposing repository credentials. Hat tip to all the devs, this is a really welcome tool!!

**Comment by spaghettifunk** [comment]
Documentation here: https://keel.sh/docs/#iam-role

**Comment by jcardoso-bv** [comment]
Is PR #438 part of `0.15.0-rc2`? I only ask as we're still unable to get Keel to pull images cross-account from our ECR repos despite all tests within the pod confirming our Kiam roles are working correctly.

**Comment by rusenask** [comment]
Yes, `0.15.0-rc2` has the latest code from master, @spaghettifunk any ideas?

**Comment by jcardoso-bv** [comment]
Very odd. Keel works fine when on the same account as our ECR but in a different account given cross-account access via ECR policies and a Kiam role it always spits out:

```
time="2019-09-10T11:24:08Z" level=error msg="trigger.poll.RepositoryWatcher.addJob: failed to get image digest" error="Get https://xxx.dkr.ecr.eu-west-1.amazonaws.com/v2/xxx/xxx-xxx-service/manifests/1.21.2: http: non-successful response (status=401 body=\"Not Authorized\\n\")" image="xxx/xxx-xxx-service:1.21.2" password= username=
time="2019-09-10T11:24:08Z" level=error msg="trigger.poll.RepositoryWatcher.Watch: failed to add image watch job" error="Get https://xxx.dkr.ecr.eu-west-1.amazonaws.com/v2/xxx/xxx-xxx-service/manifests/1.21.2: http: non-successful response (status=401 body=\"Not Authorized\\n\")" image="namespace:xxx,image:xxx.dkr.ecr.eu-west-1.amazonaws.com/xxx/xxx-xxx-service:1.21.2,provider:kubernetes,trigger:poll,sched:@every 5m,secrets:[]"
```

**Comment by spaghettifunk** [comment]
that's interesting error but I have slight feeling that it's a separate issue. I think that for your case the way the AWS object is instantiate it, needs to change. Right now what the component does is to check if there are some credentials or IAM Roles. For cross-account, I think it is necessary to keep the `session` and `config` objects of multiple accounts. 
Honestly I've never had such a situation but after some googling, I came across this [blogpost](https://maori.geek.nz/assuming-roles-in-aws-with-go-aeeb28fab418) that says the above.

I would suggest to open a new issue about this particular problem and maybe I can spend some time to fix the issue on keel side. Although, I think that there will be some changes in the configuration file (and also in the helm chart) to understand when there are `cross account` situations. This is just a brain-fart and probably I'm wrong, but at the moment I'm not sure how to solve the problem quickly.

What do you think?

**Comment by jcardoso-bv** [comment]
Sadly my understanding of the Go AWS SDK is limited at best but I think I get the gist of the issue. :) Happy enough to open up a separate issue as requested.
